bytes: bazelize iobuf_fuzz

IoannisRP opened the pull request:
Implements:
[CORE-7353](https://redpandadata.atlassian.net/browse/CORE-7353) : `iobuf_fuzz.cc`

Added extra global linking argument `-stdlib=libc++`:
without this, `libfuzzer.a`, which is included by adding `-fsanitizer=fuzzer`, introduces a dependency on `stdlibc++.so`. This ends up with both `libc++.a` and `stdlibc++.so` being linked.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none


[CORE-7353]: https://redpandadata.atlassian.net/browse/CORE-7353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

oleiman commented on src/v/bytes/tests/BUILD:
```diff
@@ -76,3 +76,18 @@ redpanda_cc_btest(
     ],
 )
+
+redpanda_cc_fuzz_test(
+    name = "iobuf_fuzz",
+    timeout = "short",
+    srcs = ["iobuf_fuzz.cc"],
+    args = [
+        "-max_total_time=30",
+        "-rss_limit_mb=8192",
+    ],
+    deps = [
+        "//src/v/base",
+        "//src/v/bytes:iobuf",
+        "//src/v/bytes:scattered_message",
+    ],
+)
```
> We don't limit to -O0

The release profile also doesn't include debug info. With optimizations and no symbolic traces, presumably any errors surfaced by a would be pretty hard to make sense of.

Is ubsan likely to catch errors in an optimized build that was missed in an unoptimized one? I'd sort of expect the result of a full optimization pass to be _less_ recognizable to the tooling, but idk.

IoannisRP replied:
> With optimizations ~and no symbolic traces~, presumably any errors would be pretty hard to make sense of.

Probably you are right, but it would be better to know that there is something potentially wrong. It might lead to painful debugging but i would argue it is better than not knowing at all. :man_shrugging:

IoannisRP replied:
> I'd sort of expect the result of a full optimization pass to be less recognizable to the tooling, but idk.

honestly, i don't know either.
My only counter-arguments are "we should be testing the code that we release" and
the quote

> It might be interesting to run w.g. with -O1 and -O2 due to different inlining. 
Running -O0 is rarely beneficial and is very slow. 

From this question
https://groups.google.com/g/libfuzzer/c/L-RQkOaJpoQ

oleiman replied:
Yeah makes sense. Particularly the bit about inlining. I wonder why we only build these in Debug in vtools... maybe @dotnwat can comment?

IoannisRP replied:
>TODO: figure out why the fuzz test is failing in the ci after the fuzzer has finished testing. It appears to be some hiccup while seastar is doing cleanup, but i need to look further into it.

I tried on a dummy fuzz test that does nothing. Just by linking against seastar, a similar error occurs. @dotnwat , could it be this was the reason it was set as debug-only?

dotnwat replied:
> I wonder why we only build these in Debug in vtools... maybe @dotnwat can comment?

The reason fuzz test is only built in debug is because that is the only build that contains sanitizers. If you build the fuzz tests with release, then you only get feedback on things like segfault.

IoannisRP commented on bazel/test.bzl:
```diff
@@ -130,6 +130,53 @@ def _redpanda_cc_test(
+        ],
+        env = env,
+        data = data,
+        cxxopts = [
+            "-g",
```
@oleiman debug info is explicitly being turned on for the fuzz target

oleiman replied:
oh nice. missed that 👍

rockwotj replied:
Do we need a `--config` for this so that `-g` is applied globally instead of only to this target? We need all libraries we link against to be `-g` too right?

IoannisRP replied:
oh... yes. Of course we do :sweat_smile:

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--8c51045a-ae81-4543-bb03-526b4c9b30eb-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/58823#01936a1b-2de9-459f-8ce3-822b59e31b05:
```
"rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_creating_and_listing_migrations"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/58823-->
<!--5afcb35e-3bab-4636-936b-4ebb3858faa0-->
<!--title-->
## Retry command for [Build#58823](https://buildkite.com/redpanda/redpanda/builds/58823)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations
```
<!--/body-->

IoannisRP commented:
- Added a global config to add debug symbols which can be used with any other configuration.

TODO: figure out why the fuzz test is failing in the ci after the fuzzer has finished testing. It appears to be some hiccup while seastar is doing cleanup, but i need to look further into it.

IoannisRP commented:
- added bazel to denote optimized builds
- disable fuzz testing for optimized builds

IoannisRP commented on bazel/BUILD:
```diff
@@ -0,0 +1,13 @@
+load("@bazel_skylib//rules:common_settings.bzl", "bool_flag")
+
+bool_flag(
+    name = "mark_optimized",
+    build_setting_default = False,
+)
+
+config_setting(
+    name = "optimized_build",
+    flag_values = {
+        ":mark_optimized": "true",
+    },
+)
```
This feels a little heavy handed. I mainly don't like that i has to be manually set when any optimization flags are passed in.

However, i couldn't find any elegant way to detect optimization flags. If you know of a nicer way, do let me know.

rockwotj replied:
Hrm, I think I would rather explicitly opt into running the fuzz tests, and disable fuzz tests otherwise.

rockwotj replied:
Otherwise you can check for `--compliation_mode` via:

```
config_setting(
    name = "opt_build",
    values = {
        "compilation_mode": "opt",
    },
)
```

IoannisRP replied:
it's also the sanitizer that add `-O1`. This also fails.

IoannisRP replied:
> I think I would rather explicitly opt into running the fuzz tests

this makes sense. The only issue is that it is not clear when they should/could be run.
Should we just leave a comment on the fuzz rule saying that it only works with unoptimized builds?

rockwotj replied:
Which is why I'd rather have a special `--config=fuzz` that enables the exact set of compilation flags required for fuzz testing, then we have some global flag that all fuzz tests can check for - similar to what you have today but instead of `mark_optimized` just do `fuzzing_enabled` or something and we have a .bazelrc configuration for making sure all the build flags work out.

I'm also confused why `-O1` fails because our current CMake debug config uses `-O1` IIRC

IoannisRP replied:
the debug itself, yes, but all the fuzz targets add an "-O0" on top. So it's a mix :sweat_smile: 

https://github.com/redpanda-data/redpanda/blob/8d90667f3d05cc694c1f9e3d423ef4b95a26f0b6/src/v/bytes/tests/CMakeLists.txt#L19-L29

IoannisRP replied:
I am pretty sure that i tested the `config:sanitizers` in the past and it was not working. However this is not what i am seeing now. I re-tested it and it works as expected.

I changed fuzz testing to only be enabled with the `config:sanitizers`.

IoannisRP commented:
- rebased to `dev`

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--0c4f6774-8a71-40cd-bfe9-3b4d62781bda-->
<!--RETRIED FAILURES-->
<!--54349ddc-2471-46e0-9bb7-45e04aca7296-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59479#0193abca-0bb2-4908-99e8-d20c23b7f220
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59479#0193abce-7f4f-40f4-9c80-649c648b9487
<!--/body-->

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2bfa43d6624d404c0ff3e076973b7a8024ae4a4f..08ec1972fb271deccba816c5b935558e32afe1a2):
- Changed `marked_optimized` to `has_sanitizers`
- Changed fuzz testing to only be enabled when `has_sanitizers` is set
- Enabled `has_sanitisers` in `config::sanitizers`

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/08ec1972fb271deccba816c5b935558e32afe1a2..4bcef5c219aaa5b0bb8db631e3027df9710af054):
- rebase to `dev`

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/4bcef5c219aaa5b0bb8db631e3027df9710af054..eeed5a3883536353622d7a585be7d35d8ae718d4): 
- cleanup fuzz testing compilation and linking arguments. Fuzz testing is only available when sanitizers are enabled. It should use whichever sanitizers the config:sanitizers is declaring.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59725-->
<!--12b2566d-a275-4fc1-a85d-ebdab7e2a2fc-->
<!--title-->
## Retry command for [Build#59725](https://buildkite.com/redpanda/redpanda/builds/59725)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5979d682-ef66-4c5b-afae-07631feee667-->
<!--TEST RESULTS-->
<!--67d98b1b-f3f4-4d2d-b582-34b7281a3316-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59725">build#59725</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c063-9217-4a67-924c-9b653551189b | FLAKY | 3/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c068-79dc-41d8-8eed-e1eb9f2507d1 | FLAKY | 3/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c068-79dc-41d8-8eed-e1eb9f2507d1 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c068-79dc-41d8-8eed-e1eb9f2507d1 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c068-79dc-41d8-8eed-e1eb9f2507d1 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59725#0193c068-79dc-41d8-8eed-e1eb9f2507d1 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59955">build#59955</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59955#0193def9-f8b6-44ec-a6f5-07f2d659e85f | FAIL          | 0/6      |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59955#0193def9-f8b7-4ff8-9eff-67e074ff5ba5 | FLAKY         | 2/6      |
</details>
<!--/body-->

rockwotj commented on .bazelrc:
```diff
@@ -111,6 +113,12 @@ build:coverage --experimental_fetch_all_coverage_outputs
+# =================================
+# Debug Symbols
+# =================================
+build:debug-symbols --copt -g
```
Can we remove this now?

IoannisRP replied:
yeah, sorry. It was a left-over from before.
Done

rockwotj reviewed (commented):
LGTM other than I think we can remove the debug symbol thing

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59955-->
<!--6ca3e855-c6c5-409b-bc38-817107aeeb4e-->
<!--title-->
## Retry command for [Build#59955](https://buildkite.com/redpanda/redpanda/builds/59955)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

IoannisRP commented:
CI failures:
- [CORE-8573](https://redpandadata.atlassian.net/browse/CORE-8573)

[CORE-8573]: https://redpandadata.atlassian.net/browse/CORE-8573?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
